quickwit/quickwit-indexing/src/source/queue_sources/processor.rs

```diff
@@ -0,0 +1,464 @@
+    /// messages until they are dropped.
+    fn spawn_visibility_tasks(
+        &mut self,
+        messages: Vec<CheckpointedMessage>,
+    ) -> Vec<(CheckpointedMessage, VisibilityTaskHandle)> {
+        messages
+            .into_iter()
+            .map(|message| {
+                let handle = spawn_visibility_task(
+                    self.queue.clone(),
+                    message.content.metadata.ack_id.clone(),
+                    message.content.metadata.initial_deadline,
+                    self.publish_lock.clone(),
+                );
+                (message, handle)
+            })
+            .collect()
+    }
+
+    /// Polls messages from the queue and prepares them for processing
+    async fn poll_messages(&mut self, ctx: &SourceContext) -> anyhow::Result<()> {
+        // receive() typically uses long polling so it can be long to respond
+        let raw_messages = ctx.protect_future(self.queue.receive()).await?;
+
+        if raw_messages.is_empty() {
+            self.observable_state.num_consecutive_empty_batches += 1;
+            return Ok(());
+        } else {
+            self.observable_state.num_consecutive_empty_batches = 0;
+        }
+
+        let preprocessed_messages = raw_messages
+            .into_iter()
+            .map(|m| m.pre_process(self.queue_params.message_type))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let categorized_using_local_state =
+            self.local_state.filter_completed(preprocessed_messages);
+
+        let categorized_using_shared_state = self
+            .shared_state
+            .checkpoint_messages(
+                &self.publish_token,
+                categorized_using_local_state.processable,
+            )
+            .await?;
+
+        // Drop visibility tasks for messages that have been processed by another pipeline
+        let completed_visibility_tasks = categorized_using_shared_state
+            .already_processed
+            .iter()
+            .filter_map(|m| self.local_state.mark_completed(m.partition_id()))
+            .collect();
+        acknowledge(&*self.queue, completed_visibility_tasks).await?;
+
+        // Acknowledge messages that have been processed by another pipeline
+        let mut already_processed = categorized_using_local_state.already_processed;
+        already_processed.extend(categorized_using_shared_state.already_processed);
+        acknowledge(&*self.queue, already_processed).await?;
+
+        let ready_messages =
+            self.spawn_visibility_tasks(categorized_using_shared_state.processable);
+
+        self.local_state.set_ready_messages(ready_messages);
+
+        Ok(())
+    }
+
+    pub async fn emit_batches(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<Duration, ActorExitStatus> {
+        if let Some(msg) = self.local_state.in_progress_mut() {
```

Comment:
I think this flow is not a great idea, also I agree this is tricky.

I would have preferred to see a stream with different tasks feeding into a channel:

```
- get message task
- download task
```

This method would then return whatever is available is in the channel.